### ISSUE TITLE
Add a spam filtering system

### ISSUE THREAD

**Issue by CounterPillow** [nyaadevs/nyaa#351]
This is a feature request for a spam filter functionality, and a discussion about how it should work.

Filters could take the form of regular expressions that are case insensitive. They can apply to a multitude of things:
* Torrent titles
* Torrent information fields
* Torrent descriptions
* Filenames
    - This would allow us to reject certain torrents like ones with BitComet padding files
* Comments

The resulting action could also be configurable, e.g.
* Reject torrent
* Reject torrent and IP-ban uploader
* Whitelist from other filters

If this was extended to include an additional parameter which restricts the filter to a certain category, we could do things like disallow executables in categories that shouldn't have executables.

**Comment by TheAMM** [comment]
#342 works as a good foothold.
I'd prefer the rules/actions be configurable outside `config.py`, but allowing very flexible rules/actions would be a pain with "proper" tables and relationships, so I'm suggesting one table with a JSON serialization of a validator(set)/action(set) combo. Adding a nice UI would require UI work, though, but maybe we can just go with a textfield for json input or something for now.

I'm thinking of stuff like
```json
{
  "name" : "Executables in wrong category",
  "validators" : [
    {"type" : "all", "validators" : [
        { "type" : "regex", "target": "filelist", "value": "\\.exe$" },
        { "type" : "value_not", "target": "torrent_category", "value": "x_xx" }
      ]
    }
  ],
  "actions" : ["reject"]
}
```
... or something. I'm certainly over-engineering this because it's neat to do.

**Comment by CounterPillow** [comment]
Okay, more armchair overengineering:

- allow filters to only apply to Anonymous uploads
- allow filters to only apply to fresh accounts (either based on time or number of torrents? Bikeshedding needed)
- action to reject with a custom message (for cases where accidentally getting caught in the spam filter is more likely)
- action to reject with a ban (for very clear cases of abuse)
- action to permit but set a cooldown on the IP (to e.g. prevent Anon floods of .exe or archives)

So, who's gonna bite the bullet and try to implement this mess?

**Comment by TheAMM** [comment]
The current implementation can handle anons, fresh accounts (requires adding a User.age or allowing `eval()`s on the filters [obviously with sane envs]), rejections & bans with message.
Cooldown will require keeping track of all uploading IPs and their last date, which will reach outside the filter logic, so it's bit of a pain.

The question is who are allowed to adjust the filters. `eval()` will make it much easier to handle some checks (eg. `len(files) == 1`) but malicious evals are still possible. Probably allow mods create filters but require admin approval, who knows.

**Comment by S-ed** [comment]
It's just my amateur opinion:
1. Why not let to only registered users upload? (even as Anonymous)
2. Make first few posts per account - premoderated

**Comment by the-eater** [comment]
Hi,

I'm going to spend coming week(s?) to implement such system. I have a general idea how I want it to look like, my current syntax proposal, that's easy to the eye, not hard to parse and easy to moderate is:

```yaml
refuse lossy:
  if:
     - not match(some,glob) torrent.files:
       - '*.flac'
       - '*.wav'
     - match(part) category: 'Lossless'
  then:
    refuse: 'Please only post lossless files in the lossless category'

refuse executables:
  if:
    - match(some,glob) torrent.files: '*.exe'
    - not match category: 'executables'
  then:
    refuse: 'Please only post executables in the executable category'
```

Feel free to give any feedback or unsolicited advice :)

**Comment by ReimuHakurei** [comment]
We do not want any false-positives with any spam filtering system.

Optimally, a spam filter will not completely block any uploads, but instead merely hide them from the index and flag them for moderator screening.

Integration of GeoIP functionality would also be useful.

Additionally, just the files contained is just one way spam could be filtered. Most of the spammers seem to follow a few patterns with their descriptions, so being able to filter based on certain suspicious patterns there would be useful.

**Comment by the-eater** [comment]
Sounds good, I guess we can then just go with a rule score system, and trigger the redflag system above a certain score

Do you have a set of spam torrent entries for me?

So I can set up some boilerplate rules.

The idea is that the rule system is pretty much abstract (python wise), so it would be easy to add additional rules.

**Comment by junh1024** [comment]
@the-eater 

```
refuse lossy:
  if:
     - not match(some,glob) [torrent.files]:
       - '*.flac'
       - '*.wav'
     - match(part) [category]: 'Lossless'
  then:
    refuse: 'Please only post lossless files in the lossless category'
```

You'll need a lot more than that m8. Also include tak tta shn dtshd thd iso wv ape , etc 

Also, perhaps not an outright refusal, but send an automated report if criteria isn't met. 

ALSO, making Lossy appear before Lossless in category might help a bit.

**Comment by S-ed** [comment]
Agree with @junh1024 
Those are lossless audio file extensions:
`.alac .flac .wv .tta .ape .shn`
And those could be lossless:
`.m4a .caf .aiff .wav .wma`

**Comment by CounterPillow** [comment]
ogg can contain FLAC though.

But this is irrelevant to the implementation of the spam filter system, and only relevant to the actual filters. In reality, disallowing .aac and .mp3 in the "Lossless" category would cover 99.9% of bases.

**Comment by S-ed** [comment]
@CounterPillow Ok, Was not aware of that. Always thought OGG just for Vorbis.

**Comment by sebastienvercammen** [comment]
Tip: Don't invent new languages/structures/syntaxes when you can just define filters as functions and use them like middleware.

Or classes if you want it in a more structured way, in pseudocode:
```
class LinkSpamFilter:
    filter(request):
        ...do something here, and return True/False to continue or reject the request
        ...or tag the request w/ a hidden tag for shadowbans, e.g.:
        if len(request.text.extractLinks()) > X:
            request.tag('potential link spam')
        return True
```
Or you can use chaining, similar to Node.js' middleware w/ `next()`.

I haven't contributed to nyaa.si yet, so apologies for the vague example that doesn't use your own code.

There's a small overhead on function calls, but this should be negligible. And everyone can just write filters in the programming language they're already familiar with. No need for new things that will never be as expansive as the general-purpose programming language you're already using.

**Comment by CounterPillow** [comment]
@sebastienvercammen that is a pretty good suggestion, thanks.
